### ISSUE TITLE
Is the base pass initialising sub-layer internals at all the right default?

### ISSUE THREAD

**Issue by KennethGrace** [latticedynamics/lumen#10]
Split out of #8, which fixed the symptom and left this.

`Stack` exists for the depth-scaled rescale. That is its irreducible job — the
scale depends on how many blocks write to the stream, so it genuinely cannot
live lower down. **Redrawing every `nn.Linear` in the trunk at a flat `init_std`
is a separate policy that arrived alongside it**, inherited from the incumbent's
`self.apply(self._init)`, and it has never been compared against anything.

Two reasons it is not obviously neutral.

A flat `0.02` ignores fan-in. `nn.Linear`'s own default does not — it scales
with `1/sqrt(fan_in)`. So the base pass replaces a fan-in-aware init with a
fan-in-blind one across every projection in the trunk, at every width, and
nobody chose that per-projection.

And #8's fix narrows the failure without eliminating the class.
`apply_init_structure()` is opt-in, so the next sub-layer with structural init
that does not implement it is clobbered exactly as silently as `state_gated`
was. There is no way to detect "this weight had structure" without knowing
*which* structure — which is precisely what made the skip-contract shape
tempting there.

**Severity is consumer-dependent and the gradient is steep.** For gradient
training an init-scale difference of this kind largely washes out. For a
population-based consumer the initial weights *are* the prior the search is
drawn from, so this is not a starting point but the distribution being searched.

Not a proposal to remove it. `CLAUDE.md`: do not canonise an untested default —
and de-canonising one on taste is the same error. What would settle it is the
comparison: matched arms under base-pass-on against base-pass-off, on the same
seeds.

Whatever lands, it is a release with a migration and not a patch — removing the
base pass changes every initial weight in every trunk.

**Comment by KennethGrace** [comment]
Ran the comparison this issue asks for. Matched arms, base pass on against base
pass off, same seed sequence and identical data, with the arms evaluated as `G`
parameter sets in a single `vmap` call — which is only possible because the state
containers are now pytree nodes (#12).

Measured on one machine — 2x Tesla P40, fp32, `torch 2.9.1`. Evidence, not a
specification.

## What the base pass is actually decisive over

The depth pass overwrites the base pass on every projection it touches, so the
base pass is only the last word on the linears it does not. Over a 4-layer trunk
(GDN mixer, Undertow local, SwiGLU MLP):

- 12 residual writes, redrawn by the depth pass — **identical in both arms by
  construction**, and width-independent (0.00707 at 4 layers)
- 48 input-side projections, where the base pass decides

**Every one of those 48 has `fan_in = d_model` exactly.** Not a spread of
fan-ins to reason about per-layer — one number per model width.

That is structural, not a property of the config it was measured on: the
governed set is exactly the projections that *read the residual stream*, so
their fan-in is the stream width by construction. Checked across `expand_k` ∈
{1, 2}, `expand_v` ∈ {1, 2}, and with the MLP present and absent — the governed
fan-in is `{d_model}` in every case, while it is the *depth* pass whose fan-ins
move (`{d_model, 2·d_model, 4·d_model}`).

## The flat draw is a width-dependent rescale nobody chose

`nn.Linear.reset_parameters` is `kaiming_uniform_(w, a=sqrt(5))`, giving bound
`1/sqrt(fan_in)` and therefore std `1/sqrt(3·fan_in)`. Against a flat
`init_std`:

```
ratio = init_std · sqrt(3 · d_model)      equal at  d_model = 1/(3·init_std²)
```

**833 for `init_std=0.02`.**

| d_model | 128 | 256 | 512 | 832 | 1024 | 2048 | 4096 |
|---|---|---|---|---|---|---|---|
| flat/default | 0.392 | 0.554 | 0.784 | 0.999 | 1.109 | 1.568 | 2.217 |

So the base pass is not fan-in-blind in a way that lands somewhere neutral. It
is **exactly equivalent to a fan-in-aware init with gain
`init_std·sqrt(3·d_model)`** — a gain curve that rises with width, crosses 1 at
833, and that nobody selected.

## Matched arms at init

`G=16` draws per arm. `spread` = across-draw std of the trunk output over its
RMS. The final RMS norm pins output scale to 1.0000 in both arms at every width,
so this is diversity of **direction** — which is what a selection step can see.

| d_model | spread on | spread off | ratio | predicted | grad ratio |
|---|---|---|---|---|---|
| 128 | 0.0267 | 0.0901 | 0.297 | 0.392 | 0.879 |
| 256 | 0.0570 | 0.1271 | 0.448 | 0.554 | 0.826 |
| 512 | 0.1260 | 0.1784 | 0.706 | 0.784 | 0.849 |
| 832 | 0.2251 | 0.2253 | **0.999** | **0.999** | **0.999** |
| 1024 | 0.2883 | 0.2482 | 1.162 | 1.109 | 1.115 |
| 2048 † | 0.5935 | 0.3336 | 1.779 | 1.568 | 1.761 |

† `G=8` — sixteen draws at that width exceed a 24GB card on parameters alone.

**At `d_model=832` the arms are indistinguishable in every metric** — spread
0.2251 against 0.2253, gradient norm 1.387e-1 against 1.388e-1, identical
activation profile through depth.

**That coincidence is the result.** When the two scales agree, the two arms
agree — so the base pass has no effect beyond the fan-in scale. No hidden
benefit from being a normal draw rather than a uniform one, nothing structural.
It is the rescale and nothing else.

Activation RMS stays within a few percent of 1.0 in both arms up to 1024. At
2048 that stops being true of one of them: base-pass-on reaches 1.267 by the
fourth block against 1.065 for base-pass-off. Above the crossover the base pass
is not only widening the prior, it is inflating the residual stream through
depth.

### One hypothesis, tested and wrong

The measured ratio is consistently more extreme than the weight-scale ratio
(0.297 against 0.392; 1.779 against 1.568). The obvious explanation is
compounding through depth. **It is wrong**, and worth recording as wrong because
it is the one to reach for:

| n_layers | 1 | 2 | 4 | 8 |
|---|---|---|---|---|
| spread ratio | 0.298 | 0.297 | 0.297 | 0.297 |
| grad ratio | 0.962 | 0.931 | 0.879 | 0.799 |

Spread is **depth-independent** — the whole gap is already there at one layer.
The cause is that a block is not linear in its input-side projections: the
associative memory writes an outer product of two of them and reads it with a
third, and SwiGLU multiplies two. Scaling by `r` moves the residual increment by
a mixture of `r`, `r²` and `r³`; the implied exponent across the five widths off
the crossover is 1.28–1.45, mean 1.36.

So the mismatch matters **more than its scale ratio suggests, at every width,
and for a one-layer trunk as much as a deep one.** Gradient norm is the
opposite — near-neutral at one layer, worsening monotonically with depth. The
two should not be quoted with the same caveat.

## The gradient half

Associative recall, loss on the final position. `vocab=16`, `pairs=4`, 4 layers,
AdamW at `lr=1e-3`, 800 steps, 5 seeds per arm. Embedding and head are drawn
under their own fixed seed *after* the trunk, so the arms' different consumption
of randomness cannot leak into them. Speed is first crossing of `loss < 0.5`,
since both arms saturate and what an init can move is the rate, not the ceiling.

| d_model | arm | steps to 0.5 | final accuracy |
|---|---|---|---|
| 128 | base-pass-on | **240 ± 18** | 0.995 ± 0.004 |
| 128 | base-pass-off | 286 ± 29 | 0.995 ± 0.004 |
| 1024 | base-pass-on | 240 ± 20 | 0.989 ± 0.004 |
| 1024 | base-pass-off | 227 ± 21 | 0.994 ± 0.007 |

Welch at 128: `off − on = +45.8` steps, `t = 3.06`, `df = 6.7`. At 1024, `−13.2`
steps, `t = −1.02` — nothing.

**This issue is right about the ceiling and wrong about the rate, at small
width.** Final accuracy is identical to within noise in all four cells, so the
difference does wash out of *where* training ends up. It does not wash out of how
long it takes: below the crossover the base pass reaches threshold about 19%
sooner, across five seeds.

## The two consumers want opposite things

At `d_model=128` the base pass, from the same rescale:

- **narrows the prior by 3.4x** — strictly worse for a population consumer,
  which has only the spread to select on
- **converges 19% sooner** — strictly better for a gradient consumer

So there may be no single right value, and `init_std` may be a knob whose
correct setting depends on which consumer is holding the trunk. That is a
sharper version of the tension this issue named, and the opposite of what a
consolidation library wants to hear.

## What is settled, and what is not

Settled: the base pass is a width-dependent rescale of 48 projections, its
entire measurable effect is that rescale, and **the direction of the effect flips
at `d_model ≈ 833`.** There is no single statement "the base pass is too small"
or "too large" — it depends where you sit relative to the crossover. A lineage
that tuned `0.02` at one width has been carrying that width's answer to every
other width since.

Not settled, and this issue's own rule holds: whether fan-in-aware is *better*.
A wider prior can just be a worse one. Nothing here licenses removing the base
pass. What it licenses is refusing to call a width-independent `0.02` neutral.

Open:

- **The rate effect is entangled with the learning rate.** AdamW partially
  normalises gradient scale, so "19% sooner" is a statement about this `lr`, not
  about optimisation in general. An `lr` sweep would separate them and has not
  been run.
- One task, one depth, one threshold, five seeds. The direction is consistent;
  the magnitude should not be quoted precisely.
- The population half is measured **at init only**. Whether a narrower initial
  population actually costs a search anything is a fitness question, and needs a
  task and a search rather than an init statistic.
- The depth pass still divides by the inherited `_WRITES_PER_BLOCK = 2` while
  these blocks make three residual writes each. That is #6; it is wrong in both
  arms identically and cannot have confounded any of the above.
